File: java_projects/glue_code/joda-money/src/main/java/org/joda/money/format/AmountPrinterParser.java
package org.joda.money.format;

import java.io.IOException;
import java.io.Serializable;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;
import org.joda.money.BigMoney;
import org.joda.money.ContextInitializer;
import org.joda.money.ExceptionHandler;

final class AmountPrinterParser implements MoneyPrinter, MoneyParser, Serializable {
  private static final long serialVersionUID = 1L;
  private static Value clz =
      ContextInitializer.getPythonClass("/format/AmountPrinterParser.py", "AmountPrinterParser");
  private Value obj;

  public AmountPrinterParser(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  public String toString() {
    //
    // return "${amount}";
    //

    // TODO: Check the type mapping below!
    return this.obj.invokeMember("toString").as(String.class);
  }

  public void parse(MoneyParseContext context) {
    //
    // final int len = context.getTextLength();
    // final MoneyAmountStyle activeStyle = style.localize(context.getLocale());
    // char[] buf = new char[len - context.getIndex()];
    // int bufPos = 0;
    // boolean dpSeen = false;
    // int pos = context.getIndex();
    // if (pos < len) {
    // char ch = context.getText().charAt(pos++);
    // if (ch == activeStyle.getNegativeSignCharacter()) {
    // buf[bufPos++] = '-';
    // } else if (ch == activeStyle.getPositiveSignCharacter()) {
    // buf[bufPos++] = '+';
    // } else if (ch >= activeStyle.getZeroCharacter()
    // && ch < activeStyle.getZeroCharacter() + 10) {
    // buf[bufPos++] = (char) ('0' + ch - activeStyle.getZeroCharacter());
    // } else if (ch == activeStyle.getDecimalPointCharacter()) {
    // buf[bufPos++] = '.';
    // dpSeen = true;
    // } else {
    // context.setError();
    // return;
    // }
    // }
    // boolean lastWasGroup = false;
    // for (; pos < len; pos++) {
    // char ch = context.getText().charAt(pos);
    // if (ch >= activeStyle.getZeroCharacter() && ch < activeStyle.getZeroCharacter() + 10) {
    // buf[bufPos++] = (char) ('0' + ch - activeStyle.getZeroCharacter());
    // lastWasGroup = false;
    // } else if (ch == activeStyle.getDecimalPointCharacter() && dpSeen == false) {
    // buf[bufPos++] = '.';
    // dpSeen = true;
    // lastWasGroup = false;
    // } else if (ch == activeStyle.getGroupingCharacter() && lastWasGroup == false) {
    // lastWasGroup = true;
    // } else {
    // break;
    // }
    // }
    // if (lastWasGroup) {
    // pos--;
    // }
    // try {
    // context.setAmount(new BigDecimal(buf, 0, bufPos));
    // context.setIndex(pos);
    // } catch (NumberFormatException ex) {
    // context.setError();
    // }
    //

    this.obj.invokeMember("parse", context);
  }

  public void print(MoneyPrintContext context, Appendable appendable, BigMoney money)
      throws IOException {
    try {
      //
      // MoneyAmountStyle activeStyle = style.localize(context.getLocale());
      // String str;
      // if (money.isNegative()) {
      // if (!activeStyle.isAbsValue()) {
      // appendable.append(activeStyle.getNegativeSignCharacter());
      // }
      // str = money.negated().getAmount().toPlainString();
      // } else {
      // str = money.getAmount().toPlainString();
      // }
      // char zeroChar = activeStyle.getZeroCharacter();
      // if (zeroChar != '0') {
      // int diff = zeroChar - '0';
      // StringBuilder zeroConvert = new StringBuilder(str);
      // for (int i = 0; i < str.length(); i++) {
      // char ch = str.charAt(i);
      // if (ch >= '0' && ch <= '9') {
      // zeroConvert.setCharAt(i, (char) (ch + diff));
      // }
      // }
      // str = zeroConvert.toString();
      // }
      // final int decPoint = str.indexOf('.');
      // final int afterDecPoint = decPoint + 1;
      // if (activeStyle.getGroupingStyle() == GroupingStyle.NONE) {
      // if (decPoint < 0) {
      // appendable.append(str);
      // if (activeStyle.isForcedDecimalPoint()) {
      // appendable.append(activeStyle.getDecimalPointCharacter());
      // }
      // } else {
      // appendable
      // .append(str.subSequence(0, decPoint))
      // .append(activeStyle.getDecimalPointCharacter())
      // .append(str.substring(afterDecPoint));
      // }
      // } else {
      // int groupingSize = activeStyle.getGroupingSize();
      // int extendedGroupingSize = activeStyle.getExtendedGroupingSize();
      // extendedGroupingSize = extendedGroupingSize == 0 ? groupingSize : extendedGroupingSize;
      // char groupingChar = activeStyle.getGroupingCharacter();
      // int pre = (decPoint < 0 ? str.length() : decPoint);
      // int post = (decPoint < 0 ? 0 : str.length() - decPoint - 1);
      // appendable.append(str.charAt(0));
      // for (int i = 1; i < pre; i++) {
      // if (isPreGroupingPoint(pre - i, groupingSize, extendedGroupingSize)) {
      // appendable.append(groupingChar);
      // }
      // appendable.append(str.charAt(i));
      // }
      // if (decPoint >= 0 || activeStyle.isForcedDecimalPoint()) {
      // appendable.append(activeStyle.getDecimalPointCharacter());
      // }
      // if (activeStyle.getGroupingStyle() == GroupingStyle.BEFORE_DECIMAL_POINT) {
      // if (decPoint >= 0) {
      // appendable.append(str.substring(afterDecPoint));
      // }
      // } else {
      // for (int i = 0; i < post; i++) {
      // appendable.append(str.charAt(i + afterDecPoint));
      // if (isPostGroupingPoint(i, post, groupingSize, extendedGroupingSize)) {
      // appendable.append(groupingChar);
      // }
      // }
      // }
      // }
      //

      this.obj.invokeMember("print", context, appendable, money);
    } catch (PolyglotException e) {
      // TODO: Handle IOException
      throw (IOException) ExceptionHandler.handle(e, "AmountPrinterParser.print");
    }
  }

  private boolean isPostGroupingPoint(int i, int post, int groupingSize, int extendedGroupingSize) {
    //
    // boolean atEnd = (i + 1) >= post;
    // if (i > groupingSize) {
    // return (i - groupingSize) % extendedGroupingSize == (extendedGroupingSize - 1)
    // && !atEnd;
    // }
    // return i % groupingSize == (groupingSize - 1) && !atEnd;
    //

    // TODO: Check the type mapping below!
    return this.obj
        .invokeMember("isPostGroupingPoint", i, post, groupingSize, extendedGroupingSize)
        .as(boolean.class);
  }

  private boolean isPreGroupingPoint(int remaining, int groupingSize, int extendedGroupingSize) {
    //
    // if (remaining >= groupingSize + extendedGroupingSize) {
    // return (remaining - groupingSize) % extendedGroupingSize == 0;
    // }
    // return remaining % groupingSize == 0;
    //

    // TODO: Check the type mapping below!
    return this.obj
        .invokeMember("isPreGroupingPoint", remaining, groupingSize, extendedGroupingSize)
        .as(boolean.class);
  }

  AmountPrinterParser(MoneyAmountStyle style) {
    //
    // this.style = style;
    //

    this.obj = clz.invokeMember("__init__", style);
  }
}

File: data/recomposed_projects/deepseek-coder-33b-instruct/joda-money/src/main/org/joda/money/format/AmountPrinterParser.py
from __future__ import annotations
from io import IOBase
import io
import os
from src.main.org.joda.money.BigMoney import *
from src.main.org.joda.money.format.GroupingStyle import *
from src.main.org.joda.money.format.MoneyAmountStyle import *
from src.main.org.joda.money.format.MoneyParseContext import *
from src.main.org.joda.money.format.MoneyParser import *
from src.main.org.joda.money.format.MoneyPrintContext import *
from src.main.org.joda.money.format.MoneyPrinter import *


class AmountPrinterParser(MoneyParser, MoneyPrinter):

    __style: MoneyAmountStyle = None

    __serialVersionUID: int = 1

    def toString(self) -> str:

        return "${amount}"

    def parse(self, context: MoneyParseContext) -> None:

        len = context.getTextLength()
        activeStyle = self.__style.localize(context.getLocale())
        buf = [""] * (len - context.getIndex())
        bufPos = 0
        dpSeen = False
        pos = context.getIndex()
        if pos < len:
            ch = context.getText().charAt(pos)
            pos += 1
            if ch == activeStyle.getNegativeSignCharacter():
                buf[bufPos] = "-"
                bufPos += 1
            elif ch == activeStyle.getPositiveSignCharacter():
                buf[bufPos] = "+"
                bufPos += 1
            elif (
                activeStyle.getZeroCharacter()
                <= ch
                < activeStyle.getZeroCharacter() + 10
            ):
                buf[bufPos] = chr(ord("0") + ch - activeStyle.getZeroCharacter())
                bufPos += 1
            elif ch == activeStyle.getDecimalPointCharacter():
                buf[bufPos] = "."
                bufPos += 1
                dpSeen = True
            else:
                context.setError()
                return
        lastWasGroup = False
        for _ in range(pos, len):
            ch = context.getText().charAt(_)
            if (
                activeStyle.getZeroCharacter()
                <= ch
                < activeStyle.getZeroCharacter() + 10
            ):
                buf[bufPos] = chr(ord("0") + ch - activeStyle.getZeroCharacter())
                bufPos += 1
                lastWasGroup = False
            elif ch == activeStyle.getDecimalPointCharacter() and not dpSeen:
                buf[bufPos] = "."
                bufPos += 1
                dpSeen = True
                lastWasGroup = False
            elif ch == activeStyle.getGroupingCharacter() and not lastWasGroup:
                lastWasGroup = True
            else:
                break
        if lastWasGroup:
            pos -= 1
        try:
            context.setAmount(BigDecimal("".join(buf[:bufPos])))
            context.setIndex(pos)
        except NumberFormatException:
            context.setError()

    def print(
        self, context: MoneyPrintContext, appendable: io.TextIOBase, money: BigMoney
    ) -> None:

        activeStyle = self.__style.localize(context.getLocale())
        str = ""
        if money.isNegative():
            if not activeStyle.isAbsValue():
                appendable.write(activeStyle.getNegativeSignCharacter())
            str = str(money.negated().getAmount())
        else:
            str = str(money.getAmount())

        zeroChar = activeStyle.getZeroCharacter()
        if zeroChar != "0":
            diff = ord(zeroChar) - ord("0")
            zeroConvert = list(str)
            for i in range(len(str)):
                ch = str[i]
                if ch >= "0" and ch <= "9":
                    zeroConvert[i] = chr(ord(ch) + diff)
            str = "".join(zeroConvert)

        decPoint = str.find(".")
        afterDecPoint = decPoint + 1
        if activeStyle.getGroupingStyle() == GroupingStyle.NONE:
            if decPoint < 0:
                appendable.write(str)
                if activeStyle.isForcedDecimalPoint():
                    appendable.write(activeStyle.getDecimalPointCharacter())
            else:
                appendable.write(str[:decPoint])
                appendable.write(activeStyle.getDecimalPointCharacter())
                appendable.write(str[afterDecPoint:])
        else:
            groupingSize = activeStyle.getGroupingSize()
            extendedGroupingSize = activeStyle.getExtendedGroupingSize()
            extendedGroupingSize = (
                extendedGroupingSize if extendedGroupingSize != 0 else groupingSize
            )
            groupingChar = activeStyle.getGroupingCharacter()
            pre = len(str) if decPoint < 0 else decPoint
            post = len(str) - decPoint - 1 if decPoint >= 0 else 0
            appendable.write(str[0])
            for i in range(1, pre):
                if self.__isPreGroupingPoint(
                    pre - i, groupingSize, extendedGroupingSize
                ):
                    appendable.write(groupingChar)
                appendable.write(str[i])
            if decPoint >= 0 or activeStyle.isForcedDecimalPoint():
                appendable.write(activeStyle.getDecimalPointCharacter())
            if activeStyle.getGroupingStyle() == GroupingStyle.BEFORE_DECIMAL_POINT:
                if decPoint >= 0:
                    appendable.write(str[afterDecPoint:])
            else:
                for i in range(post):
                    appendable.write(str[i + afterDecPoint])
                    if self.__isPostGroupingPoint(
                        i, post, groupingSize, extendedGroupingSize
                    ):
                        appendable.write(groupingChar)

    def __isPostGroupingPoint(
        self, i: int, post: int, groupingSize: int, extendedGroupingSize: int
    ) -> bool:

        atEnd = (i + 1) >= post
        if i > groupingSize:
            return (i - groupingSize) % extendedGroupingSize == (
                extendedGroupingSize - 1
            ) and not atEnd
        return i % groupingSize == (groupingSize - 1) and not atEnd

    def __isPreGroupingPoint(
        self, remaining: int, groupingSize: int, extendedGroupingSize: int
    ) -> bool:

        if remaining >= groupingSize + extendedGroupingSize:
            return (remaining - groupingSize) % extendedGroupingSize == 0
        return remaining % groupingSize == 0

    def __init__(self, style: MoneyAmountStyle) -> None:

        self.__style = style
